src/App.php

```diff
@@ -376,7 +376,10 @@ public function addStyle($style)
      */
     public function normalizeClassNameApp($name)
     {
-        return '\\'.__NAMESPACE__.'\\'.$name;
+        if (class_exists('\\'.__NAMESPACE__.'\\'.$name)) {
```

Comment:
__NAMESPACE__ return the namespace with single "\" you need to str_replace("\","\\",_NAMESPACE__) i think.

NOTE: I think we need to talk about declaration of argument type :
- fast no control at runtime :: in comments for static analysis
- "slow" strict check at runtime :: in function declaration raise exception

related video of Rasmus Lerdorf - https://www.youtube.com/watch?v=rKXFgWP-2xQ

php 5.6 support ended on 01-01-2019 ( https://www.php.net/supported-versions.php )

Reply:
```

        /* can be defined as constant */
        $ns_sepa = chr(92);

        /** test against untouched name */
        if (class_exists($name)) {
            return $name;
        }

        /** normalize namespace */
        $normalized_namespace = $ns_sepa.__NAMESPACE__;

        /** normalize name */
        $normalized_name = $ns_sepa.trim($name,$ns_sepa);

        /** test against normalize name */
        if (class_exists($normalized_name)) {
            return $normalized_name;
        }

        /** test against normalized namespace + normalize name */
        if (class_exists($normalized_namespace.$normalized_name)) {
            return $normalized_namespace.$normalized_name;
        }

        /** last chance to check it */
        /* give error even with this extra check
        $atk4_namespace = $ns_sepa.'atk4/ui';
        if (class_exists($atk4_namespace.$normalized_name)) {
            return $atk4_namespace.$normalized_name;
        }
        */

        /* it will throw exception for sure check to help sleepy dev */
        // char(47) = / aka wrong slash used as separator
        if(strpos($name,chr(47)) !== false)
        {
            throw new Exception('Wrong separator used for class, detect / in place of \ called from class ' . $name);
        }

        /** class not exists DIE */
        throw new Exception('Class with name ' . $name . ' not exists, checked ' . 
$normalized_name. ' | ' . $normalized_namespace . $normalized_name);
```

BUT even if in this way it works as intended, it redeclare classes and throw errors in some demos unittest.
Errors are due to misuse of wrong slash in place of backslash in $name.
We can make a str_replace and lose performance and solving a problem of the dev, in my opinion this is wrong, or throw an exception and make it changes the code or add a flag and make this function switchable in App.

Reply:
i remember somewhere there was something on this subject of usage of slash in place of backslash

https://agile-core.readthedocs.io/en/latest/factory.html#namespace

we need to find where to put the str_replace of slash to backslash, before all or at the end and make another call to the function and finally throw exception?

```
public function normalizeClassNameApp($name)
    {
        /* can be defined as constant */
        $ns_sepa = chr(92);

        /** test against untouched name */
        if (class_exists($name)) {
            return $name;
        }

        /** normalize namespace */
        $normalized_namespace = $ns_sepa.__NAMESPACE__;

        /** normalize name */
        $normalized_name = $ns_sepa.trim($name,$ns_sepa);

        /** test against normalize name */
        if (class_exists($normalized_name)) {
            return $normalized_name;
        }

        /** test against normalized namespace + normalize name */
        if (class_exists($normalized_namespace.$normalized_name)) {
            return $normalized_namespace.$normalized_name;
        }

        /** last chance to check it */
        /* give error even with this extra check
        $atk4_namespace = $ns_sepa.'atk4/ui';
        if (class_exists($atk4_namespace.$normalized_name)) {
            return $atk4_namespace.$normalized_name;
        }
        */

        /* it will throw exception for sure check to help sleepy dev */
        // char(47) = / aka wrong slash used as separator
        if(strpos($name,chr(47)) !== false)
        {
            return $this->normalizeClassNameApp(str_replace(chr(47), chr(92), $name));
        }

        /** class not exists DIE */
        throw new Exception('Class with name ' . $name . ' not exists, checked ' . $normalized_name. ' | ' . $normalized_namespace . $normalized_name);
    }
```